From 4dec1b2fdb87223a3a595d8ddf47e77beff336ff Mon Sep 17 00:00:00 2001
From: Oliver Heger <oheger@apache.org>
Date: Wed, 9 Feb 2011 21:05:00 +0000
Subject: [PATCH] [CONFIGURATION-434] Allow comment characters to be part of
 property values. To be recognized as a comment character, the character must
 be preceded by whitespace.

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/configuration/trunk@1069097 13f79535-47bb-0310-9956-ffa450edef68
---
 src/changes/changes.xml                       |  7 +++++++
 .../HierarchicalINIConfiguration.java         | 14 ++++++++++---
 .../TestHierarchicalINIConfiguration.java     | 20 +++++++++++++++++++
 3 files changed, 38 insertions(+), 3 deletions(-)

diff --git a/src/changes/changes.xml b/src/changes/changes.xml
index 4814d0b7f2..5bf283fd8c 100644
--- a/src/changes/changes.xml
+++ b/src/changes/changes.xml
@@ -23,6 +23,13 @@
 
   <body>
     <release version="1.7" date="in SVN" description="">
+      <action dev="oheger" type="fix" issue="CONFIGURATION-434">
+        HierarchicalINIConfiguration now recognizes comment characters in
+        property definitions only if they are preceded by whitespace. Thus
+        comment characters can now be part of the property value. This is for
+        instance required for the definition of file paths which use the
+        semicolon as path separator.
+      </action>
       <action dev="oheger" type="fix" issue="CONFIGURATION-433">
         Minor improvements of the support for indexed properties in
         ConfigurationDynaBean.
diff --git a/src/java/org/apache/commons/configuration/HierarchicalINIConfiguration.java b/src/java/org/apache/commons/configuration/HierarchicalINIConfiguration.java
index 9687694334..763e30d390 100644
--- a/src/java/org/apache/commons/configuration/HierarchicalINIConfiguration.java
+++ b/src/java/org/apache/commons/configuration/HierarchicalINIConfiguration.java
@@ -398,6 +398,12 @@ public void load(Reader reader) throws ConfigurationException
      * <pre>
      * 'value' ; comment -&gt; value
      * </pre>
+     * Note that a comment character is only recognized if there is at least one
+     * whitespace character before it. So it can appear in the property value,
+     * e.g.:
+     * <pre>
+     * C:\\Windows;C:\\Windows\\system32
+     * </pre>
      *
      * @param val the value to be parsed
      * @param reader the reader (needed if multiple lines have to be read)
@@ -420,6 +426,7 @@ private static String parseValue(String val, BufferedReader reader) throws IOExc
             int i = quoted ? 1 : 0;
 
             StringBuffer result = new StringBuffer();
+            char lastChar = 0;
             while (i < value.length() && !stop)
             {
                 char c = value.charAt(i);
@@ -452,17 +459,18 @@ else if (escape && quote == c)
                 }
                 else
                 {
-                    if (!isCommentChar(c))
+                    if (isCommentChar(c) && Character.isWhitespace(lastChar))
                     {
-                        result.append(c);
+                        stop = true;
                     }
                     else
                     {
-                        stop = true;
+                        result.append(c);
                     }
                 }
 
                 i++;
+                lastChar = c;
             }
 
             String v = result.toString();
diff --git a/src/test/org/apache/commons/configuration/TestHierarchicalINIConfiguration.java b/src/test/org/apache/commons/configuration/TestHierarchicalINIConfiguration.java
index eb76bd65e9..27ae0adeac 100644
--- a/src/test/org/apache/commons/configuration/TestHierarchicalINIConfiguration.java
+++ b/src/test/org/apache/commons/configuration/TestHierarchicalINIConfiguration.java
@@ -683,6 +683,26 @@ public void testSaveKeysWithDelimiters() throws ConfigurationException
         assertEquals("Wrong value (2)", "test2", conf.getString(section + ".test2"));
     }
 
+    /**
+     * Tests whether a value which contains a semicolon can be loaded
+     * successfully. This test is related to CONFIGURATION-434.
+     */
+    public void testValueWithSemicolon() throws ConfigurationException
+    {
+        final String path =
+                "C:\\Program Files\\jar\\manage.jar;"
+                        + "C:\\Program Files\\jar\\guiLauncher.jar";
+        final String content =
+                "[Environment]" + LINE_SEPARATOR + "Application Type=any"
+                        + LINE_SEPARATOR + "Class Path=" + path + "  ;comment"
+                        + LINE_SEPARATOR + "Path=" + path
+                        + "\t; another comment";
+        HierarchicalINIConfiguration config = setUpConfig(content);
+        assertEquals("Wrong class path", path,
+                config.getString("Environment.Class Path"));
+        assertEquals("Wrong path", path, config.getString("Environment.Path"));
+    }
+
     /**
      * A thread class for testing concurrent access to the global section.
      */
